rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -82,12 +88,95 @@ impl HyperlaneChain for CosmosProvider {
 
 #[async_trait]
 impl HyperlaneProvider for CosmosProvider {
-    async fn get_block_by_hash(&self, _hash: &H256) -> ChainResult<BlockInfo> {
-        todo!() // FIXME
+    async fn get_block_by_hash(&self, hash: &H256) -> ChainResult<BlockInfo> {
+        let tendermint_hash = Hash::from_bytes(Algorithm::Sha256, hash.as_bytes())
+            .expect("block hash should be of correct size");
+
+        // TODO add proper error handling
+        let response = self
+            .rpc_client
+            .block_by_hash(tendermint_hash)
+            .await
+            .map_err(|_| ChainCommunicationError::from_other_str("generic error"))?;
+
+        let received_hash = H256::from_slice(response.block_id.hash.as_bytes());
+
+        if &received_hash != hash {
+            return Err(ChainCommunicationError::from_other_str(
+                "received incorrect block",
+            ));
+        }
+
+        let block = response
+            .block
+            .ok_or_else(|| ChainCommunicationError::from_other_str("empty block info"))?;
+
+        let time: OffsetDateTime = block.header.time.into();
+
+        let block_info = BlockInfo {
+            hash: hash.to_owned(),
+            timestamp: time.unix_timestamp() as u64,
+            number: block.header.height.value(),
+        };
+
+        Ok(block_info)
     }
 
-    async fn get_txn_by_hash(&self, _hash: &H256) -> ChainResult<TxnInfo> {
-        todo!() // FIXME
+    async fn get_txn_by_hash(&self, hash: &H256) -> ChainResult<TxnInfo> {
+        let tendermint_hash = Hash::from_bytes(Algorithm::Sha256, hash.as_bytes())
+            .expect("transaction hash should be of correct size");
+
+        // TODO add proper error handling
+        let response = self
+            .rpc_client
+            .tx(tendermint_hash, true)
+            .await
+            .map_err(|_| ChainCommunicationError::from_other_str("generic error"))?;
+
+        let received_hash = H256::from_slice(response.hash.as_bytes());
+
+        if &received_hash != hash {
+            return Err(ChainCommunicationError::from_other_str(
+                "received incorrect transaction",
+            ));
+        }
+
+        let tx = Tx::from_bytes(&response.tx).map_err(|_| {
```

Comment:
might also be nice to propagate the error, share the tx hash if it's useful for debugging

Reply:
I'll do error handling in future PRs.